src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/executor/SchemaCache.java

```diff
@@ -1,93 +1,313 @@
+  /**
+   * Gets a listener to use with the {@link CqlSessionFactory} to remove the schema cache entries
+   * when the DB sends schema change events.
+   */
+  public SchemaChangeListener getSchemaChangeListener() {
+    return new SchemaCacheSchemaChangeListener(this);
+  }
+
+  /**
+   * Gets a consumer to use with the {@link CQLSessionCache} to remove the schema cache entries when
+   * a tenant is deactivated.
+   */
+  public CQLSessionCache.DeactivatedTenantConsumer getDeactivatedTenantConsumer() {
+    return new SchemaCacheDeactivatedTenantConsumer(this);
+  }
 
+  /** Gets or loads the schema object for the given namespace and collection or table name. */
   public Uni<SchemaObject> getSchemaObject(
-      RequestContext dataApiRequestInfo,
-      Optional<String> tenant,
+      RequestContext requestContext,
       String namespace,
       String collectionName,
       boolean forceRefresh) {
 
-    // TODO: refactor, this has duplicate code, the only special handling the OSS has is the tenant
-    // check
+    Objects.requireNonNull(namespace, "namespace must not be null");
+
+    // Aaron 1 may 2025 - Based on existing logic I thought we should have a non null tenantId in
+    // all cases other than
+    // when using cassandra as a backend.
+    var resolvedTenantId =
+        (databaseType == DatabaseType.CASSANDRA)
+            ? requestContext.getTenantId().orElse(DEFAULT_TENANT)
+            : requestContext.getTenantId().orElseThrow();
 
-    if (CASSANDRA.equals(operationsConfig.databaseConfig().type())) {
-      // default_tenant is for oss run
-      // TODO: move the string to a constant or config, why does this still check the tenant if this
-      // is for OSS ?
-      final NamespaceCache namespaceCache =
-          schemaCache.get(
-              new CacheKey(Optional.of(tenant.orElse("default_tenant")), namespace),
-              this::addNamespaceCache);
-      return namespaceCache.getSchemaObject(dataApiRequestInfo, collectionName, forceRefresh);
+    var tableBasedSchemaCache =
+        keyspaceCache.get(new KeyspaceCacheKey(resolvedTenantId, namespace));
+    Objects.requireNonNull(
+        tableBasedSchemaCache, "keyspaceCache must not return null tableBasedSchemaCache");
+    return tableBasedSchemaCache.getSchemaObject(requestContext, collectionName, forceRefresh);
+  }
+
+  private TableBasedSchemaCache onLoad(SchemaCache.KeyspaceCacheKey key) {
+    if (LOGGER.isTraceEnabled()) {
+      LOGGER.trace("onLoad() - tenantId: {}, keyspace: {}", key.tenantId(), key.keyspace());
     }
 
-    final NamespaceCache namespaceCache =
-        schemaCache.get(new CacheKey(tenant, namespace), this::addNamespaceCache);
-    return namespaceCache.getSchemaObject(dataApiRequestInfo, collectionName, forceRefresh);
+    // MUST not use the session cache in the ctor, OK to create a query executor for each inner
+    // cache here
+    var queryExecutor = new QueryExecutor(sessionCacheFactory.create(), operationsConfig);
```

Comment:
Now realizing "SessionCacheFactory" is slightly misleading as it only creates and shares a single(ton) instance. So it's almost as it should be called "SessionCacheProvider" or such.
Just because here it might read like we create separate cache for each executor (which we don't).

Reply:
Will rename to SessionCacheProvider after reviewing all the notes.